lightning/src/routing/network_graph.rs

```diff
@@ -243,6 +288,380 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		}
+
+		// We filter the short_channel_ids to those inside the query range.
+		// The most significant 3-bytes of the short_channel_id are the block.
+		{
+			let mut filtered_short_channel_ids: Vec<u64> = msg.short_channel_ids.clone().into_iter().filter(|short_channel_id| {
+				let block = short_channel_id >> 40;
+				return block >= query_range_tasks.get(their_node_id).unwrap().first_blocknum as u64 && block <= task_last_block as u64;
+			}).collect();
+			let task = query_range_tasks.get_mut(their_node_id).unwrap();
+			task.short_channel_ids.append(&mut filtered_short_channel_ids);
+		}
+
+		// The final message is indicated by a last_blocknum that is equal to
+		// or greater than the query's last_blocknum.
+		if msg_last_block >= task_last_block {
+			log_debug!(self.logger, "Completed query_channel_range: peer={}, first_blocknum={}, number_of_blocks={}", log_pubkey!(their_node_id), task_first_blocknum, task_number_of_blocks);
+
+			// We can now fire off a query to obtain routing messages for the
+			// accumulated short_channel_ids.
+			{
+				let task = query_range_tasks.get_mut(their_node_id).unwrap();
+				let mut short_channel_ids = Vec::new();
+				std::mem::swap(&mut short_channel_ids, &mut task.short_channel_ids);
+				self.query_short_channel_ids(their_node_id, task.chain_hash, short_channel_ids)?;
+			}
+
+			// We can remove the query range task now that the query is complete.
+			query_range_tasks.remove(their_node_id);
+		}
+		Ok(())
+	}
+
+	fn handle_reply_short_channel_ids_end(&self, their_node_id: &PublicKey, msg: &ReplyShortChannelIdsEnd) -> Result<(), LightningError> {
+		log_debug!(self.logger, "Handling reply_short_channel_ids_end peer={}, full_information={}", log_pubkey!(their_node_id), msg.full_information);
+
+		// First we obtain a lock on the task hashmap. In order to avoid borrowing issues
+		// we will access the task as needed.
+		let mut query_short_channel_ids_tasks = self.query_short_channel_ids_tasks.lock().unwrap();
+
+		// If there is no existing task then we have received an unknown
+		// message and should return an error.
+		if query_short_channel_ids_tasks.get(their_node_id).is_none() {
+			return Err(LightningError {
+				err: String::from("Unknown reply_short_channel_ids_end message"),
+				action: ErrorAction::IgnoreError,
+			});
+		}
+
+		// If the reply's chain_hash does not match the task's chain_hash then
+		// the reply is malformed and we should return an error.
+		if msg.chain_hash != query_short_channel_ids_tasks.get(their_node_id).unwrap().chain_hash {
+			query_short_channel_ids_tasks.remove(their_node_id);
+			return Err(LightningError {
+				err: String::from("Received reply_short_channel_ids_end with incorrect chain_hash"),
+				action: ErrorAction::IgnoreError
+			});
+		}
+
+		// If the remote node does not have up-to-date information for the
+		// chain_hash they will set full_information=false. We can fail
+		// the result and try again with a different peer.
```

Comment:
AFAIU this protocol, lack of information fullness doesn't imply you've not received any `channel_announcement`/`channel_update`. Function comment should warn initial caller to update it set of scids before to proceed with another peer.